src/object.d

```diff
@@ -155,15 +164,269 @@ auto opEquals(Object lhs, Object rhs)
     return lhs.opEquals(rhs) && rhs.opEquals(lhs);
```

Comment:
When object.opEquals is removed, how do you ensure that this doesn't result in a recursive call to this template again?

Reply:
It's been a long time, but I looked over this again and I think this could still result in a recursive call.
```D
class Foo
{
    bool opEquals(int x);
}
```

In this case, `Foo` has an `opEquals` member, however, if you call `foo1.opEquals(foo2)`, it will recursively call `opEquals` here.  I'm not sure the best way to fix this...maybe something like `__traits(getMember, lhs, "opEquals)(rhs)`?

Reply:
Update.  I tried to see if this would in fact result in a recursive call and it didn't seem to.  My guess is that if an object contains the member `opEquals`, and you call `obj.opEquals`, then it will ALWAYS call the member function even if the arguments are invalid (it will never fall back to calling a UFCS method).  Can anyone confirm this?  I attempted to find this in the spec but was unsuccessful.

Reply:
It is my understanding that UFCS is only ever attempted if there is no member function with that name. The [spec](https://dlang.org/spec/function.html#pseudo-member) implies that when it says

> When UFCS rewrite is necessary, compiler searches the name on accessible module level scope, in order from the innermost scope.

but it isn't very specific about it. There was a complaint about it recently in D.Learn, because someone wanted to overload a range's `front` with a `front` function that took arguments, and it didn't work. So, all the evidence that I've seen seems to indicate that UFCS is only attempted when it _needs_ to be attempted, which presumably reduces symbol look-up times (since it only has to look at the member functions if there's one with that name), and it avoids hijacking when someone screws up calling a member function.

Reply:
Created an issue to clarify this behavior in the spec https://issues.dlang.org/show_bug.cgi?id=18523   

So long as this is the correct behavior, the change in the PR looks correct.